apm-agent-plugins/apm-javalin-plugin/src/main/java/co/elastic/apm/agent/javalin/JavalinInstrumentation.java

```diff
@@ -0,0 +1,189 @@
+                            );
+                            logger.debug("This Javalin version is supported");
+                        } catch (NoSuchMethodException e) {
+                            logger.info("The current Javalin version is not supported, only 3.13.8+ versions are supported");
+                            handlerTypeMethodHandle = NOOP;
+                        } catch (Throwable throwable) {
+                            logger.error("Cannot get a method handle for io.javalin.http.Context#handlerType(), Javalin will not be traced", throwable);
+                            handlerTypeMethodHandle = NOOP;
+                        }
+                    }
+                }
+            }
+            HandlerType handlerType = null;
+            if (handlerTypeMethodHandle != NOOP) {
+                try {
+                    //noinspection ConstantConditions
+                    handlerType = (HandlerType) handlerTypeMethodHandle.invoke(context);
+                } catch (Throwable throwable) {
+                    logger.error("Cannot determine Javalin HandlerType. Javalin cannot be traced");
+                }
+            }
+            return handlerType;
+        }
+
+        @Nullable
+        @Advice.OnMethodEnter(suppress = Throwable.class, inline = false)
+        public static Object setSpanAndTransactionName(@Advice.This Handler handler,
+                                                       @Advice.Argument(0) Context ctx) {
+
+            HandlerType handlerType = getHandlerType(ctx);
+
+            // exit early if instrumentation cannot work because of missing Context.handlerType() method due to older version
+            if (handlerType == null) {
+                return null;
+            }
+
+            final Transaction transaction = tracer.currentTransaction();
+            if (transaction == null) {
+                return null;
+            }
+
+            final String handlerClassName = handler.getClass().getName();
+
+            // do not create an own span for JavalinServlet.addHandler, as this not added by the users code and leads to confusion
+            if (handlerClassName.startsWith("io.javalin.http.JavalinServlet")) {
+                return null;
+            }
+
+            // transaction name gets only set if we are dealing with a HTTP method processing, not before/after handlers
+            if (handlerType.isHttpMethod()) {
+                final StringBuilder name = transaction.getAndOverrideName(PRIO_HIGH_LEVEL_FRAMEWORK, false);
+                if (name != null) {
+                    transaction.setFrameworkName(FRAMEWORK_NAME);
+                    transaction.setFrameworkVersion(VersionUtils.getVersion(Handler.class, "io.javalin", "javalin"));
+                    transaction.withType("request");
+
+                    name.append(handlerType.name()).append(" ").append(ctx.endpointHandlerPath());
```

Comment:
I figured that this means it will only called once within the HTTP handler instead of within any before/after HTTP handlers as well, thus this placement